WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+     * is consistent with the other platforms and allows for the possibility of setting a non-Wikipedia domain
+     * like "commons.wikimedia.org" and "wikidata.org" for multimedia/metadata-related in-app analytics.
+     *
+     * Cases where instrumentation would set a `domain`:
+     * - reading or editing an article
+     * - managing watchlist
+     * - interacting with feed
+     * - searching
+     *
+     * Cases where it might not be necessary for instrumentation to set a `domain`:
+     * - changing settings
+     * - managing reading lists
+     * - navigating map of nearby articles
+     * - multi-lingual features like Suggested Edits
+     * - marking session start/end; in which case schema and `data` should have a `languages` field
+     *   where user's list of languages can be stored, although it might make sense to set it to the domain
+     *   associated with the user's 1st preferred language – in which case use
+     *   `MWKLanguageLinkController.sharedInstance().appLanguage?.siteURL()!.host!`
+     */
+    public func log(stream: String, schema: String, data: [String: Any], domain: String? = nil) -> Void {
+        if !ENABLED {
+            return
+        }
+        var meta: [String: String]
+        if data.keys.contains("meta") {
+            meta = data["meta"]! as! [String: String]
+        } else {
+            meta = [String: String]()
+        }
+        if domain != nil {
+            meta["domain"] = domain!
+        }
+        var data = data
+        data["meta"] = meta
+
+        /*
+         * The top-level field `client_dt` is for recording the time the event
+         * was generated. EventGate sets `meta.dt` during ingestion, so for
+         * analytics events that field is used as "timestamp of reception" and
+         * is used for partitioning the events in the database. See Phab:T240460
+         * for more information.
+         */
+        if !data.keys.contains("client_dt") {
+            LAST_TS = Date()
+            data["client_dt"] = ISO8601_FORMATTER.string(from: LAST_TS)
+        }
+
+        /*
+         * Generated events have the session ID attached to them before stream
+         * config is available (in case they're generated offline) and before
+         * they're cc'd to any other streams (once config is available).
+         */
+        if !data.keys.contains("session_id") {
+            data["session_id"] = session_id()
+        }
+
+        if CONFIG == nil {
+            input_buffer.append((stream, schema, data, domain))
+            return
+        } else {
+            // CC to other streams, even if this stream does not exist
+            if COPIED.keys.contains(stream) {
+                for cc_stream in COPIED[stream]! {
+                    log(stream: cc_stream, schema: schema, data: data, domain: domain)
+                }
+            }
+            if !(CONFIG!.keys.contains(stream)) {
+                return
+            }
+        }
```

Comment:
This snippet is reworked with guard, it's generally preferable to keep things from nesting too much. Also safely unwrapping the copiedStream to avoid the `!`:

```
        guard let config = config else {
            let logItem = LogItem(stream: stream, schema: schema, data: data, domain: domain)
            inputBuffer.append(logItem)
            return
        }

        // CC to other streams, even if this stream does not exist
        if let copiedStream = copied[stream] {
            for ccStream in copiedStream {
                log(stream: ccStream, schema: schema, data: data, domain: domain)
            }
        }
        
        if !(config.keys.contains(stream)) {
            return
        }
```

Reply:
I can't thank you enough for all this great educational feedback.